File: python/Q11_container_with_most_water.py
#
# @lc app=leetcode id=11 lang=python3
#
# [11] Container With Most Water
#


# @lc code=start
class Solution:
    def maxArea(self, height: list[int]) -> int:
        left: int = 0
        right: int = len(height) - 1
        max_area: int = 0

        while left < right:
            max_area = max(max_area, (right - left) * min(height[left], height[right]))

            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return max_area


# @lc code=end

File: rust/src/leetcode/Q11_container_with_most_water.rs
/*
 * @lc app=leetcode id=11 lang=rust
 *
 * [11] Container With Most Water
 */

// @lc code=start
use std::cmp::{max, min};

impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut result: i32 = 0;

        let mut left: usize = 0;
        let mut right: usize = height.len() - 1;

        while left < right {
            let current: i32 = (right - left) as i32 * min(height[left], height[right]);

            if current > result {
                result = current
            }

            if height[left] < height[right] {
                left += 1;
            } else {
                right -= 1;
            }
        }

        result
    }
}
// @lc code=end
